### PR TITLE
Feat/authentication/auth persistence

### PR DESCRIPTION
**Pull request by Harrish92**
# Auth Persistence

## Description
This PR introduces authentication persistence using Supabase. It closes issue #291 . 

One edge case that is not addressed in this PR is the correctness of auth persistence for sign-up. The authentication persistence redirects the user to the profile screen page even when they haven't finished creating their profile.

New states are introduced in this PR, namely `SuccessIsLoggedIn` and `SuccessLogOut`. These special cases of `Success` state represent respectively the case when the user is already successfully logged in and the case when the user successfully logs out. The unit tests are updated for the new states.

A logic function for the user authentication state is introduced in main. When the user is in `SuccessIsLoggedIn` state, it should redirect them to the profile screen . Otherwise it does nothing.

Persistence for Google authentication is reserved for another possible issue.

## Changes
- Implemented functionality to persist user authentication state using Supabase.
- Ensured that users remain logged in across app restarts.
- Added a mechanism to check if a user is currently authenticated.

#### Added
- `app/src/test/java/com/android/periodpals/MainActivityTest.kt`

#### Modified
- `app/src/main/java/com/android/periodpals/MainActivity.kt`
- `app/src/main/java/com/android/periodpals/model/authentication/AuthenticationViewModel.kt`
- `app/src/main/java/com/android/periodpals/model/user/UserAuthenticationState.kt`
- `app/src/test/java/com/android/periodpals/model/authentication/AuthenticationViewModelTest.kt`

#### Removed
- None

## Dependencies Added
- None

## Testing
- Modified unit tests for authentication persistence functions.
- Added unit tests for user authentication state logic.

### CODE REVIEW

File: app/src/main/java/com/android/periodpals/MainActivity.kt [hunk 1]
```diff
@@ -143,6 +154,8 @@ fun PeriodPalsApp(
 
   val locationViewModel: LocationViewModel = viewModel(factory = LocationViewModel.Factory)
 
+  userAuthStateLogic(authenticationViewModel, navigationActions)
```
Comment (agonzalez-r) [inline app/src/main/java/com/android/periodpals/MainActivity.kt:157]: good thinking doing a separate function for this. This separation improves readability, maintainability, and testability of the code.

File: app/src/main/java/com/android/periodpals/model/user/UserAuthenticationState.kt [hunk 2]
```diff
@@ -23,6 +23,22 @@ sealed class UserAuthenticationState {
    */
   data class Success(val message: String) : UserAuthenticationState()
 
+  /**
+   * Represents the success state of the authentication process. Indicates that the user is logged
+   * in.
+   *
+   * @property message A message with additional information about the successful login.
+   */
+  data class SuccessIsLoggedIn(val message: String) : UserAuthenticationState()
```
Comment (agonzalez-r) [inline app/src/main/java/com/android/periodpals/model/user/UserAuthenticationState.kt:32]: consistent implementation, good job!

File: app/src/test/java/com/android/periodpals/MainActivityTest.kt [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+@OptIn(ExperimentalCoroutinesApi::class)
+class MainActivityTest {
+
+  @Mock private lateinit var authModel: AuthenticationModelSupabase
+  private lateinit var authenticationViewModel: AuthenticationViewModel
+
+  private lateinit var navigationActions: NavigationActions
+
+  @ExperimentalCoroutinesApi @get:Rule var mainCoroutineRule = MainCoroutineRule()
```
Comment (agonzalez-r) [inline app/src/test/java/com/android/periodpals/MainActivityTest.kt:28]: you can remove `@ExperimentalCoroutinesApi `annotation here as you already applied it at the class level

File: app/src/test/java/com/android/periodpals/MainActivityTest.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  @Before
+  fun setUp() {
+    MockitoAnnotations.openMocks(this)
+    authenticationViewModel = AuthenticationViewModel(authModel)
+    navigationActions = mock(NavigationActions::class.java)
+  }
+
+  @Test
+  fun userAuthStateLogicIsUserLoggedInTest() = runBlocking {
```
Comment (agonzalez-r) [inline app/src/test/java/com/android/periodpals/MainActivityTest.kt:38]: I know sometimes we insist on using camelcase for test naming for consistency, but in this case, this name is incomprehensible, i suggest using snakecase and naming it something like ``when user is logged in, navigate to profile`()`. Idem for the other test

File: app/src/test/java/com/android/periodpals/MainActivityTest.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    doAnswer { inv -> inv.getArgument<() -> Unit>(0)() }
+        .`when`(authModel)
+        .isUserLoggedIn(any<() -> Unit>(), any<(Exception) -> Unit>())
+
+    authenticationViewModel.isUserLoggedIn()
+
+    userAuthStateLogic(authenticationViewModel, navigationActions)
+
```
Comment (agonzalez-r) [inline app/src/test/java/com/android/periodpals/MainActivityTest.kt:46]: Could you add an assert to verify the value of userAuthState? like 
` assertEquals( UserAuthenticationState.SuccessIsLoggedIn,authenticationViewModel.userAuthenticationState.value )`

File: app/src/test/java/com/android/periodpals/MainActivityTest.kt [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  @Test
+  fun userAuthStateLogicIsUserNotLoggedInTest() = runBlocking {
+    doAnswer { inv -> inv.getArgument<(Exception) -> Unit>(1)(Exception("user not logged in")) }
+        .`when`(authModel)
+        .isUserLoggedIn(any<() -> Unit>(), any<(Exception) -> Unit>())
+
+    authenticationViewModel.isUserLoggedIn()
+
+    userAuthStateLogic(authenticationViewModel, navigationActions)
+
```
Comment (agonzalez-r) [inline app/src/test/java/com/android/periodpals/MainActivityTest.kt:59]: and here: 
` assertEquals(UserAuthenticationState.Error, authenticationViewModel.userAuthenticationState.value)`

**Review by agonzalez-r — APPROVED** [review]
## Summary
Good job on such an efficient implementations. It solves one of the biggest problems we've had on these few weeks, which is that we had to login each time we opened the app. 
## Important 
#### Code Quality
- Code is clean and consistent. I just suggested some renaming for the tests for better readability. 
#### Functionality
- Implementation is very simple, it mostly uses the preexisting authentication view model, therefore the functionality is seamless.
#### Performance 
I tested manually and I couldn't find any faults. Except for the edge case that you mention in the description, which we agreed that it was a little extreme to prioritise fixing it. 
## Testing
Tests all pass. Just suggested further assertions. But overall very good coverage!
## Documentation
Missing documentation in the function you added in MainActivity. 
## Suggestions
For the future, we could maybe add more to the `MainActivityTest` you created, since there are other things to test from `MainActivity`. 
## Steps before PR approved
Consider the nitpicks and suggestions I left in the comments

File: app/src/test/java/com/android/periodpals/MainActivityTest.kt [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  @Test
+  fun navigate_to_profile_when_user_is_logged_in() = runBlocking {
+    doAnswer { inv -> inv.getArgument<() -> Unit>(0)() }
+        .`when`(authModel)
+        .isUserLoggedIn(any<() -> Unit>(), any<(Exception) -> Unit>())
+
+    authenticationViewModel.isUserLoggedIn()
+
+    userAuthStateLogic(authenticationViewModel, navigationActions)
+
```
Comment (taghizadlaura) [inline app/src/test/java/com/android/periodpals/MainActivityTest.kt:47]: Why did you resolve Ale's comment about `assertEquals(UserAuthenticationState.SuccessIsLoggedIn, authenticationViewModel.userAuthenticationState.value)` and `assertEquals(UserAuthenticationState.Error, authenticationViewModel.userAuthenticationState.value)`? I don't see any response to her request or any changes in the code. Sorry, maybe I'm a bit blind, but could you explain what you did or decided?

**Review by taghizadlaura — APPROVED** [review]
## Summary
LGTM !
This PR introduces authentication persistence using Supabase, ensuring that users remain logged in across app restarts. New states, `SuccessIsLoggedIn` and `SuccessLogOut`, are added consistently to handle authentication states. 

## Important 

#### Code Quality
LGTM :
- **Readability**: The code is clear and readable.
- **Consistency**: The changes maintain consistency with the current code structure and logic. The introduction of the new states for authentication is handled well.

#### Functionality
LGTM :
- The core functionality of persisting authentication is working well. Users are correctly kept logged in across app restarts, and the correct logic is followed when a user is in the `SuccessIsLoggedIn` state (redirecting them to the profile screen).

#### Performance 
- No performance concerns identified. 

#### Security 
- No security issues.

## Testing
- **LGTM**: The unit tests for the authentication persistence functions have been updated well, and new tests for user authentication states are in place. 

## Documentation
LGTM. The overall documentation of the changes is clear.

## Suggestions
- As you mentioned it would also be nice to implement persistence for Google authentication in a future issue.

## Steps before PR approved
LGTM, but please answer my question about the resolved comments of Ale.

File: app/src/test/java/com/android/periodpals/MainActivityTest.kt (hunk 7 above)
Comment (Harrish92) [reply app/src/test/java/com/android/periodpals/MainActivityTest.kt:47]: > Why did you resolve Ale's comment about assertEquals(UserAuthenticationState.SuccessIsLoggedIn, authenticationViewModel.userAuthenticationState.value) and assertEquals(UserAuthenticationState.Error, authenticationViewModel.userAuthenticationState.value)? I don't see any response to her request or any changes in the code. Sorry, maybe I'm a bit blind, but could you explain what you did or decided?

@taghizadlaura 
I did something equivalent with the variable `result `, because  `assertEquals` did not work.
